ARCHITECTURE.md

```diff
@@ -0,0 +1,231 @@
+1. [CLI](CLI) code. CLI code shares the `IstioControlPlaneSpec` API with
+the controller, but allows manifests to be generated and optionally applied from the command line without the need to
+run a privileged controller in the cluster. 
+1. [Migration tools](Migration_tools). The migration tools are intended to
+automate configuration migration from Helm to the operator.
+
+The operator code uses the new Helm charts in the [istio/installer](https://github.com/istio/installer) repo. It is not
+compatible with the older charts in [istio/istio](https://github.com/istio/istio/tree/master/install/kubernetes/helm).
+See the istio/installer repo for details about the new charts and why they were created. Briefly, the new charts
+are intended to support production ready deployments of Istio that follow best practises like canarying for upgrade.
+
+## Terminology
+
+Throughout the document, the following terms are used:
+
+- `IstioControlPlaneSpec`: The API directly defined in the
+[IstioControlPlaneSpec proto](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto),
+including feature and component groupings, namespaces and enablement, and per-component K8s settings. The term new API
+in this document excludes the pass-through to Helm values fields in the proto.
+- Helm values.yaml API, implicitly defined through the various values.yaml files in the
+[Helm charts](https://github.com/istio/installer) and schematized in the operator through
+[Go structs](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/values_types.go).
+
+## IstioControlPlaneSpec API
+
+The `IstioControlPlaneSpec` API is intended to replace the installation and K8s parts of Helm values.yaml.
+
+### Features and components
+
+The operator has a very similar structure to istio/installer: components are grouped into features.
+`IstioControlPlaneSpec` defines functional settings at the feature level. Functional settings are those that performs some
+function in the Istio control plane without necessarily being tied to any one component that runs in a Deployment.
+Component settings are those that necessarily refer to a particular Deployment or Service. For example, the number
+of Pilot replicas is clearly a component setting, because it refers to a component which is a Deployment in the
+cluster. Most K8s platform settings are necessarily component settings.
+The available features and the components that comprise it are as follows:
+
+| Feature | Components |
+|---------|------------|
+Base | CRDs
+Traffic Management | Pilot
+Policy | Policy
+Telemetry | Telemetry
+Security | Citadel
+Security | Node agent
+Security | Cert manager
+Configuration management | Galley
+Gateways | Ingress gateway
+Gateways | Egress gateway
+AutoInjection | Sidecar injector
+
+Features and components are defined in the
+[name](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/name/name.go#L38) package.
+
```

Comment:
One question I have at this point is whether the operator manages dependencies between features and/or between components. E.g. does it disable Pilot automatically if CRDs is not enabled?
There is no mention of dependency management in the rest of the doc, so I assume the answer is "no."
I think it would be great to add a line here, or in the Enablement section to clarify that explicitly.

Reply:
The answer is it only validates dependencies but does not automatically mutate configs. I've added some wording to that effect in the Validations section. 
For mutating the config, I think it's something we can consider for the future. I think for now it's probably clear enough that we don't do such things without making an explicit statement about it.